packages/flutter/lib/src/widgets/magnifier.dart

```diff
@@ -0,0 +1,545 @@
+  /// with its entry animation.
+  ///
+  /// To control what overlays are shown in the magnifier, utilize [below]. See
+  /// [overlayEntry] for more details on how to utilize [below].
+  ///
+  /// If the magnifier already exists (i.e. [overlayEntry] != null), then [show] will
+  /// reshow the old overlay.
+  Future<void> show({
+    required BuildContext context,
+    required WidgetBuilder builder,
+    Widget? debugRequiredFor,
+    OverlayEntry? below,
+  }) async {
+    if (overlayEntry != null) {
+      if (animationController?.status == AnimationStatus.dismissed) {
+        await animationController!.forward();
+      }
+
+      return;
+    }
+
+    final OverlayState? overlayState = Overlay.of(
+      context,
+      rootOverlay: true,
+      debugRequiredFor: debugRequiredFor,
+    );
+
+    final CapturedThemes capturedThemes = InheritedTheme.capture(
+      from: context,
+      to: Navigator.maybeOf(context)?.context,
+    );
+
+    overlayEntry = OverlayEntry(
+      builder: (BuildContext context) => capturedThemes.wrap(builder(context)),
+    );
+    overlayState!.insert(overlayEntry!, below: below);
+
+    if (animationController != null) {
+      await animationController?.forward();
+    }
+  }
+
+  /// Schedules a hide of the magnifier.
+  ///
+  /// If this [MagnifierController] has an [AnimationController],
+  /// then [hide] reverses the animation controller and waits
+  /// for the animation to complete. Then, if [removeFromOverlay]
+  /// is true, remove the magnifier from the overlay.
+  ///
+  /// In general, [removeFromOverlay] should be true, unless
+  /// the magnifier needs to preserve states between shows / hides.
+  Future<void> hide({bool removeFromOverlay = true}) async {
+    if (overlayEntry == null) {
+      return;
+    }
+
+    if (animationController != null) {
+      await animationController?.reverse();
+    }
+
+    if (removeFromOverlay) {
+      overlayEntry?.remove();
+      overlayEntry = null;
+    }
+  }
+
+  /// A utility for calculating a new [Rect] from [rect] such that
+  /// [rect] is fully constrained within [bounds].
+  ///
+  /// Any point in the output rect is guaranteed to also be a point contained in [bounds].
+  ///
+  /// It is a runtime error for [rect].width to be greater than [bounds].width,
+  /// and it is also an error for [rect].height to be greater than [bounds].height.
+  ///
+  /// This algorithm translates [rect] the shortest distance such that it is entirely within
+  /// [bounds].
+  ///
+  /// If [rect] is already within [bounds], no shift will be applied to [rect] and
+  /// [rect] will be returned as-is.
+  ///
+  /// It is perfectly valid for the output rect to have a point along the edge of the
+  /// [bounds]. If the desired output rect requires that no edges are parallel to edges
+  /// of [bounds], see [Rect.deflate] by 1 on [bounds] to achieve this effect.
+  static Rect shiftWithinBounds({
```

Comment:
How does this differ from Rect.intersect? If that doesn't serve the same purpose, then we should think about moving this out of MagnifierController, as it seems more generically about Rects.

Reply:
intersect returns the intersection of two rects, i.e. the part the overlaps. This just translates the first rect so it fully fits in the second rect. 

I guess these are the same iff intersect is called when one rect completely overlaps another. 

It is more generally about rects, but I don't see anywhere in the code after looking where someone needs something like this but me. I'll rename it to be more magnifier specific, but we should keep this in mind in the event that we see ourselves shifting rects within bounds often in the future.